### PR TITLE
feat: Modified processGlobalQueue to be able to handle results in runMethod

### PR DESCRIPTION
**Pull request by AFink**
As I was following the template for big bots, the rest proxy is an external instance of the bot stack.

If I now want to run requests and respond to my bot instance properly, I would need a way to get only JSON responses to return them to the bot via HTTP responses.

I now rewrote how processGlobalQueue returns the data to runMethod. Now, runMethod throws the errors, but the function runProxyMethod is able to return proper JSON objects to the http server. ( This file could be placed into a plugin).

For all users, nothing would change with the current implementation.

Greetings, Andreas

### CODE REVIEW

File: rest/processGlobalQueue.ts [hunk 1]
```diff
@@ -118,18 +118,33 @@ export async function processGlobalQueue(rest: RestManager) {
...
           }
 
-          request.request.reject(new Error(`[${response.status}] ${error}`));
+          request.request.reject({
+            ok: false,
+            status: response.status,
+            error: {
+              raw: new Error(`[${response.status}] ${error}`),
+              name: error,
+            },
```
Comment (itohatweb) [inline rest/processGlobalQueue.ts:133]: Completely remove `new Error()` and make the `runMethod` function construct it.

```suggestion
            error,
```

File: rest/processGlobalQueue.ts [hunk 2]
```diff
@@ -118,18 +118,33 @@ export async function processGlobalQueue(rest: RestManager) {
 
         // If NOT rate limited remove from queue
         if (response.status !== 429) {
+          let json = undefined;
           if (response.type) {
-            console.log(JSON.stringify(await response.json()));
+            json = await response.json();
+            console.log(JSON.stringify(json));
```
Comment (itohatweb) [inline rest/processGlobalQueue.ts:124]:
```suggestion
            json = JSON.stringify(await response.json());
            console.log(json);
```

File: rest/processGlobalQueue.ts [hunk 3]
```diff
@@ -118,18 +118,33 @@ export async function processGlobalQueue(rest: RestManager) {
...
 
-          request.request.reject(new Error(`[${response.status}] ${error}`));
+          request.request.reject({
+            ok: false,
+            status: response.status,
+            error: {
+              raw: new Error(`[${response.status}] ${error}`),
+              name: error,
+            },
+            body: json ? JSON.stringify(json) : undefined,
```
Comment (itohatweb) [inline rest/processGlobalQueue.ts:134]:
```suggestion
            body: json,
```

File: rest/processGlobalQueue.ts [hunk 4]
```diff
@@ -118,18 +118,33 @@ export async function processGlobalQueue(rest: RestManager) {
...
-            request.request.reject(
-              new Error(`[${response.status}] The request was rate limited and it maxed out the retries limit.`),
-            );
+            request.request.reject({
+              ok: false,
+              status: 429,
+              error: {
+                raw: new Error(`[${response.status}] The request was rate limited and it maxed out the retries limit.`),
+                name: "The request was rate limited and it maxed out the retries limit.",
+              },
```
Comment (itohatweb) [inline rest/processGlobalQueue.ts:146]: Completely remove `new Error()` and make the `runMethod` function construct it.

```suggestion
              error: "The request was rate limited and it maxed out the retries limit.",
```

File: rest/processGlobalQueue.ts [hunk 5]
```diff
@@ -118,18 +118,33 @@ export async function processGlobalQueue(rest: RestManager) {
...
         } else {
           if (request.payload.retryCount++ >= rest.maxRetryCount) {
             rest.debug(`[REST - RetriesMaxed] ${JSON.stringify(request.payload)}`);
             // REMOVE ITEM FROM QUEUE TO PREVENT RETRY
-            request.request.reject(
-              new Error(`[${response.status}] The request was rate limited and it maxed out the retries limit.`),
-            );
+            request.request.reject({
+              ok: false,
+              status: 429,
```
Comment (itohatweb) [inline rest/processGlobalQueue.ts:142]:
```suggestion
              status: response.status,
```

File: rest/processGlobalQueue.ts [hunk 6]
```diff
@@ -143,21 +158,31 @@ export async function processGlobalQueue(rest: RestManager) {
...
       if (response.status === 204) {
         rest.debug(`[REST - FetchSuccess] URL: ${request.urlToUse} | ${JSON.stringify(request.payload)}`);
-        request.request.respond({ status: 204 });
+        request.request.respond({
+          ok: true,
+          status: 204,
+        });
       } else {
         // CONVERT THE RESPONSE TO JSON
         const json = await response.json();
```
Comment (itohatweb) [inline rest/processGlobalQueue.ts:167]: This should be `JSON.stringify()` here already

File: rest/processGlobalQueue.ts [hunk 7]
```diff
@@ -143,21 +158,31 @@ export async function processGlobalQueue(rest: RestManager) {
...
           body: JSON.stringify(json),
         });
       }
     } catch (error) {
       // SOMETHING WENT WRONG, LOG AND RESPOND WITH ERROR
       rest.debug(`[REST - fetchFailed] Payload: ${JSON.stringify(request.payload)} | Error: ${error}`);
-      request.request.reject(error);
+      request.request.reject({
+        ok: false,
+        status: 500, // TODO?
```
Comment (itohatweb) [inline rest/processGlobalQueue.ts:181]: Maybe use a status code like `599` or `598` there

File: rest/processGlobalQueue.ts [hunk 8]
```diff
@@ -143,21 +158,31 @@ export async function processGlobalQueue(rest: RestManager) {
...
+        error: {
+          raw: error,
+        },
```
Comment (itohatweb) [inline rest/processGlobalQueue.ts:184]:
```suggestion
        error: "Internal Proxy Error",
```

File: rest/rest.ts [hunk 9]
```diff
@@ -49,14 +50,28 @@ export const rest = {
...
   url: string;
   method: string;
-  respond: (payload: { status: number; body?: string }) => unknown;
-  reject: (error: unknown) => unknown;
+  respond: (payload: RestRequestResponse) => unknown;
+  reject: (payload: RestRequestRejection) => unknown;
+}
+
+export interface RestRequestResponse {
+  ok: boolean; // TODO: do we even need this?
```
Comment (itohatweb) [inline rest/rest.ts:65]:
```suggestion
  ok: boolean;
```

File: rest/rest.ts [hunk 10]
```diff
@@ -49,14 +50,28 @@ export const rest = {
...
+  status: number;
+  body?: string;
+}
+
+export interface RestRequestRejection extends RestRequestResponse {
+  error: {
+    raw: Error;
+    name?: string;
+  };
```
Comment (itohatweb) [inline rest/rest.ts:74]:
```suggestion
  error: string;
```

File: rest/runMethod.ts [hunk 11]
```diff
@@ -53,12 +53,11 @@ export async function runMethod<T = any>(
       {
         url,
         method,
-        reject: (error: unknown) => {
-          errorStack.message = (error as Error)?.message;
+        reject: (data) => {
+          errorStack.message = (data.error.raw).message;
```
Comment (itohatweb) [inline rest/runMethod.ts:57]:
```suggestion
          errorStack.message = `[${data.status}] ${data.error}`;
```

File: rest/runProxyMethod.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
+import { RestManager } from "../bot.ts";
+import { SnakeCasedPropertiesDeep } from "../types/util.ts";
+import { RestRequestRejection, RestRequestResponse } from "./rest.ts";
+
+export type ProxyMethodResponse<T> = Omit<RestRequestResponse | RestRequestRejection, "body"> & { body?: T };
+
+// Left out proxy request, because it's not needed here
+// this file could also be moved to a plugin.
+export async function runProxyMethod<T = any>(
```
Comment (itohatweb) [inline rest/runProxyMethod.ts:9]: Where and how would this be used?

**Review by itohatweb — CHANGES_REQUESTED** [review]

File: rest/runProxyMethod.ts (hunk 12 above)
Comment (AFink) [reply rest/runProxyMethod.ts:9]: This function would be used in your proxy, just like here: https://github.com/discordeno/discordeno/blob/main/template/bigbot/src/rest/mod.ts#L48

File: rest/runProxyMethod.ts [hunk 13]
```diff
@@ -13,7 +13,7 @@ export async function runProxyMethod<T = any>(
   body?: unknown,
   retryCount = 0,
   bucketId?: string,
-): Promise<ProxyMethodResponse<SnakeCasedPropertiesDeep<T>>> {
+): Promise<Omit<(RestRequestResponse | RestRequestRejection), "body"> & { body?: Record<string, unknown> }> {
```
Comment (AFink) [inline rest/runProxyMethod.ts:16]: Maybe these type declarations should be changed.

Comment (itohatweb) [reply rest/runProxyMethod.ts:16]: Keep the type. If you need, then just remove the generic from it.

**Comment by itohatweb** [comment]
Looks like tests are failing.
image

File: rest/runProxyMethod.ts [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+  // No proxy so we need to handle all rate limiting and such
+  return new Promise((resolve, reject) => {
+    rest.processRequest(
+      rest,
+      {
+        url,
+        method,
+        reject: (data) => {
+          const { body: b, ...r } = data;
+          return { body: data.status !== 204 ? JSON.parse(b ?? "{}") : (undefined as unknown as T), ...r };
```
Comment (AFink) [inline rest/runProxyMethod.ts:34]: Should this be resolve or reject?

**Review by itohatweb — APPROVED** [review]

**Pull request merged**